### PR TITLE
Fix for DirichletBoundary on non-vertex nodesets.

### PR DESCRIPTION
**Pull request by dknez**
We added support for DirichletBoundary conditions based on nodesets a while ago. However, I just found out that this doesn't work in cases where the nodesets are on "non-vertex nodes" (e.g. on a mid-side node on a second order mesh).

This PR fixes it for me.

### CODE REVIEW

File: src/base/dof_map_constraints.C [hunk 1]
```diff
@@ -431,6 +431,9 @@ private:
         std::vector<bool> is_boundary_node(elem->n_nodes(), false),
           is_boundary_edge(elem->n_edges(), false),
           is_boundary_side(elem->n_sides(), false);
+
+        // We also maintain a separate list of nodeset-based boundary nodes
+        std::vector<bool> is_boundary_nodeset(elem->n_nodes(), false);
```
Comment (jwpeterson) [inline src/base/dof_map_constraints.C]: Even if there's no chance this code will run threaded (is there?) I'd still like to lodge my usual formal complaint against `std::vector<bool>`.

See section on [data races](http://www.cplusplus.com/reference/vector/vector-bool/) at the bottom.

Comment (dknez) [inline src/base/dof_map_constraints.C]: OK, but I just made it consistent with the preceding lines. Do you propose to change all four of them?

Comment (dknez) [inline src/base/dof_map_constraints.C]: P.S. If you do want to change all four lines, perhaps you could do that in a separate commit (outside of this PR)? It seems to me that is not germane to this PR.

Comment (jwpeterson) [inline src/base/dof_map_constraints.C]: > OK, but I just made it consistent with the preceding lines.

I know, just giving @roystgnr a hard time.

> P.S. If you do want to change all four lines, perhaps you could do that in a separate commit (outside of this PR)?

Agreed, sorry didn't mean to make it sound like you were supposed to do that.

Comment (roystgnr) [inline src/base/dof_map_constraints.C]: Hey, you know what else suffers from race conditions during multithreaded access?  "i++" on an integer.  Let me know when you've gone through the code and fixed all those.  :-P

**Comment by roystgnr** [comment]
Looks good to me.

**Comment by roystgnr** [comment]
:+1:

File: src/base/dof_map_constraints.C (hunk 1 above)
Comment (jwpeterson) [inline src/base/dof_map_constraints.C]: > "i++" on an integer.

The difference being that `i++` is a write operation, while *read-only* access into `vector<bool>` is thread-unsafe...

Comment (roystgnr) [inline src/base/dof_map_constraints.C]: I think read-only is still supposed to be safe (in C++11, anyway; the C++03 standard on data races says "what's a thread?"); what the standard allows vector<bool> to screw up is concurrent writes to different elements.  https://stackoverflow.com/questions/16130494/are-stdmap-and-stdvector-thread-safe

Comment (jwpeterson) [inline src/base/dof_map_constraints.C]: Thanks, didn't know about the distinction on this issue between the different standards.  I wonder if it means that operator[] access is slower for `vector<bool>` in C++11 because they have to do some locking to enforce 23.2.2.

Comment (roystgnr) [inline src/base/dof_map_constraints.C]: I don't think any locking is needed.  Basically the way that const ```vector<bool>::operator[]``` works is to bitwise-and the appropriate char against a mask.  You can do that as many times as you want concurrently and not break anything.

With non-const ```vector<bool>::operator[]```, by contrast, under the hood ```=``` is similar to ```++``` for integers: it has to first read data (that char), then modify that data (with a bitwise-or for true or bitwise-and for false), then write the data back.  If someone else tries to do the same thing at the same time, then you risk them both getting the same initial data and then only one of the two modifications gets written back.
